MetaLamp/nft-marketplace/src/Plutus/Contracts/NftMarketplace/OnChain/Core/NFT.hs

```diff
@@ -17,84 +17,41 @@
-import qualified Data.List                      as HL
-import qualified Data.Text                      as T
-import           Ext.Plutus.Contracts.Auction   (AuctionFee, AuctionParams (..))
-import qualified GHC.Generics                   as Haskell
+import           Control.Lens                           ((&), (.~), (?~), (^.))
+import qualified Control.Lens                           as Lens
+import qualified Crypto.Hash                            as Hash
+import qualified Data.Aeson                             as J
+import qualified Data.ByteArray                         as BA
+import qualified Data.List                              as HL
+import qualified Data.Text                              as T
+import qualified GHC.Generics                           as Haskell
 import           Ledger
-import qualified Ledger.Constraints             as Constraints
-import qualified Ledger.Typed.Scripts           as Scripts
-import qualified Ledger.Value                   as V
-import           Plutus.Abstract.Percentage     (Percentage)
+import qualified Ledger.Constraints                     as Constraints
+import qualified Ledger.Typed.Scripts                   as Scripts
+import qualified Ledger.Value                           as V
+import           Plutus.Abstract.Percentage             (Percentage)
 import           Plutus.Contract
 import           Plutus.Contract.StateMachine
-import qualified Plutus.Contract.StateMachine   as SM
-import qualified Plutus.Contracts.Services.Sale as Sale
+import qualified Plutus.Contract.StateMachine           as SM
+import           Plutus.Contracts.Services.Auction.Core (Auction (..),
+                                                         AuctionFee)
+import qualified Plutus.Contracts.Services.Auction.Core as Auction
+import qualified Plutus.Contracts.Services.Sale         as Sale
 import qualified PlutusTx
-import qualified PlutusTx.AssocMap              as AssocMap
-import           PlutusTx.Prelude               hiding (Semigroup (..))
-import           Prelude                        (Semigroup (..))
-import qualified Prelude                        as Haskell
--- TODO can't use POSIXTime directly because of custom JSON instances defined in Plutus:
--- generated purescript type has generic instances
-type POSIXTimeT = Integer
+import qualified PlutusTx.AssocMap                      as AssocMap
+import           PlutusTx.Prelude                       hiding (Semigroup (..))
+import           Prelude                                (Semigroup (..))
+import qualified Prelude                                as Haskell
 
 -- Category = [BuiltinByteString]
 -- 1. acts as a list of category with nested subcategories
 -- 2. acts as a list of tags
 type IpfsCid = BuiltinByteString
 type IpfsCidHash = BuiltinByteString
--- type Auction = (ThreadToken, PubKeyHash, Value, POSIXTimeT)
 type Category = [BuiltinByteString]
-type LotLink = Either Sale.Sale Auction
+type LotLink = Either Sale.Sale Auction.Auction
```

Comment:
I think using `Either` to represent a custom type that just happens to have two alternatives is a bad practice. It is not consistent, extendable, and is just confusing overall. Either has a clear bias towards `Right` value, while any custom type should not have that, unless it also a Monad, which is usually not the case. 

I suggest replacing this type as well as `PutLotRedeemer` and `RemoveLotRedeemer` with custom data.